control-plane/pkg/reconciler/channel/v2/channelv2.go

```diff
@@ -0,0 +1,624 @@
+				Ready:              corev1.ConditionFalse,
+				Message:            fmt.Sprint("Subscription not ready. ", err),
+			})
+			globalErr = multierr.Append(globalErr, err)
+		} else {
+			currentCgs = append(currentCgs, cg) //adding reconciled consumer group to map
+			if cg.IsReady() {
+				logger.Debug("marking subscription as ready")
+				channel.Status.Subscribers = append(channel.Status.Subscribers, v1.SubscriberStatus{
+					UID:                s.UID,
+					ObservedGeneration: s.Generation,
+					Ready:              corev1.ConditionTrue,
+				})
+			} else {
+				topLevelCondition := cg.GetConditionSet().Manage(cg.GetStatus()).GetTopLevelCondition()
+				if topLevelCondition == nil {
+					msg := fmt.Sprint("Subscription not ready. ", "Failed to reconcile consumer group. ", "Consumer group not ready")
+					channel.Status.Subscribers = append(channel.Status.Subscribers, v1.SubscriberStatus{
+						UID:                s.UID,
+						ObservedGeneration: s.Generation,
+						Ready:              corev1.ConditionUnknown,
+						Message:            msg,
+					})
+					//globalErr = multierr.Append(globalErr, errors.New(msg))
+				} else {
+					msg := fmt.Sprint("Subscription not ready. ", topLevelCondition.Reason, topLevelCondition.Message)
+					channel.Status.Subscribers = append(channel.Status.Subscribers, v1.SubscriberStatus{
+						UID:                s.UID,
+						ObservedGeneration: s.Generation,
+						Ready:              corev1.ConditionFalse,
+						Message:            msg,
+					})
+					//globalErr = multierr.Append(globalErr, errors.New(msg))
+				}
+			}
+		}
+	}
+
+	// Get all consumer groups associated with this Channel
+	selector := labels.SelectorFromSet(map[string]string{KafkaChannelNameLabel: channel.Name})
+	channelCgs, err := r.ConsumerGroupLister.ConsumerGroups(channel.GetNamespace()).List(selector)
+	if err != nil {
+		globalErr = multierr.Append(globalErr, err)
+	}
+	for _, cg := range channelCgs {
+		found := false
+		for _, ccg := range currentCgs {
+			if ccg.Name == cg.Name {
+				found = true
+			}
+		}
+		if !found { //ConsumerGroup needs to be deleted since it isn't associated with an existing subscriber (subscriber may have been deleted)
+			err := r.finalizeConsumerGroup(ctx, cg)
+			if err != nil {
+				globalErr = multierr.Append(globalErr, err)
+			}
+		}
+	}
+
+	return globalErr
+}
+
+func (r Reconciler) reconcileConsumerGroup(ctx context.Context, channel *messagingv1beta1.KafkaChannel, s *v1.SubscriberSpec, topicName string, bootstrapServers []string) (*internalscg.ConsumerGroup, error) {
+	newcg := &internalscg.ConsumerGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      string(s.UID),
+			Namespace: channel.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				*kmeta.NewControllerRef(channel),
+			},
+			Labels: map[string]string{
+				KafkaChannelNameLabel: channel.Name, // identifies the new ConsumerGroup as associated with this channel (same namespace)
+			},
```

Comment:
We need to make sure that this label is set and immutable in the ConsumerGroup validation.

Reply:
Currently, we only have this label for consumer groups created by KafkaChannel reconcilers. If adding as part of validation, do you recommend we add this label in KafkaSource and Trigger reconciler code too for CGs? @pierDipi

Reply:
Eh, that's right, so let's not do this, probably we can only add the immutability check.

When the label is present in the previous version, it must be present in the current version of the ConsumerGroup.

Reply:
ok - that works - will add